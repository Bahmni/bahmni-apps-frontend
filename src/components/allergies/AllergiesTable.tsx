import React, { useMemo } from 'react';
import { Tag } from '@carbon/react';
import { ExpandableDataTable } from '@components/expandableDataTable/ExpandableDataTable';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { useAllergies } from '@hooks/useAllergies';
import { formatAllergies } from '@services/allergyService';
import { FormattedAllergy } from '@types/allergy';
import { formatDateTime } from '@utils/date';
import { generateId, capitalize } from '@utils/common';

/**
 * Component to display patient allergies in a DataTable with expandable rows
 */
const AllergiesTable: React.FC = () => {
  const patientUUID = usePatientUUID();
  const { allergies, loading, error } = useAllergies(patientUUID);

  // Define table headers
  const headers = useMemo(
    () => [
      { key: 'display', header: 'Allergy' },
      { key: 'manifestation', header: 'Reaction(s)' },
      { key: 'status', header: 'Status' },
      { key: 'severity', header: 'Severity' },
      { key: 'recorder', header: 'Provider' },
      { key: 'recordedDate', header: 'Recorded Date' },
    ],
    [],
  );

  // Format allergies for display
  const formattedAllergies = useMemo(() => {
    if (!allergies || allergies.length === 0) return [];
    return formatAllergies(allergies);
  }, [allergies]);

  // Create row class names array for styling rows with severe allergies
  const rowClassNames = useMemo(() => {
    return formattedAllergies.map((allergy) =>
      allergy.reactions?.some((reaction) => reaction.severity === 'severe')
        ? 'criticalCell'
        : '',
    );
  }, [formattedAllergies]);

  // Function to render cell content based on the cell ID
  const renderCell = (allergy: FormattedAllergy, cellId: string) => {
    switch (cellId) {
      case 'display':
        return allergy.display;
      case 'status':
        return (
          <Tag type={allergy.status === 'Active' ? 'green' : 'gray'}>
            {allergy.status}
          </Tag>
        );
      case 'manifestation':
        return allergy.reactions
          ? allergy.reactions
              .map((reaction) => reaction.manifestation.join(', '))
              .join(', ')
          : 'Not available';
      case 'severity':
        return allergy.reactions
          ? allergy.reactions
              .map((reaction) => reaction.severity)
              .filter((severity): severity is string => !!severity)
              .map((severity) => capitalize(severity))
              .join(', ')
          : 'Not available';
      case 'recorder':
        return allergy.recorder || 'Not available';
      case 'recordedDate':
        return formatDateTime(allergy.recordedDate || '');
    }
  };

  // Function to render expanded content for an allergy
  const renderExpandedContent = (allergy: FormattedAllergy) => {
    if (allergy.note && allergy.note.length > 0) {
      return (
        <p style={{ padding: '0.5rem' }} key={generateId()}>
          {allergy.note.join(', ')}
        </p>
      );
    }
    return undefined;
  };

  return (
    <div
      style={{ width: '100%', paddingTop: '1rem' }}
      data-testid="allergy-table"
    >
      <ExpandableDataTable
        tableTitle="Allergies"
        rows={formattedAllergies}
        headers={headers}
        renderCell={renderCell}
        renderExpandedContent={renderExpandedContent}
        loading={loading}
        error={error}
        ariaLabel="Patient allergies"
        emptyStateMessage="No allergies found"
        rowClassNames={rowClassNames}
      />
    </div>
  );
};

export default AllergiesTable;
