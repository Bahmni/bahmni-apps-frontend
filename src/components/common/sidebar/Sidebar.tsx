import React from 'react';
import { SideNav, SideNavItems, SideNavLink } from '@carbon/react';
import BahmniIcon from '@components/common/bahmniIcon/BahmniIcon';
import { ICON_SIZE } from '@constants/icon';
import * as styles from './styles/Sidebar.module.scss';
import { useTranslation } from 'react-i18next';

/**
 * Interface defining the properties for sidebar items.
 * Each item represents a navigation option in the sidebar.
 *
 * @interface
 * @property {string} id - Unique identifier for the item
 * @property {string} icon - Icon name in FontAwesome format (e.g., "fa-clipboard-list")
 * @property {string} label - Display text for the item
 */
export interface SidebarItemProps {
  id: string;
  icon: string;
  label: string;
}

/**
 * Sidebar component that renders a vertical list of navigation items using Carbon SideNav.
 * Manages active state and click events at the component level.
 *
 * @component
 * @param {SidebarItemProps[]} items - Array of sidebar items to render
 * @param {string | null} activeItemId - ID of the currently active/selected item
 * @param {function} onItemClick - Callback function executed when an item is clicked, receives the item ID as parameter
 */
interface SidebarProps {
  items: SidebarItemProps[];
  activeItemId: string | null;
  onItemClick: (itemId: string) => void;
}

const Sidebar: React.FC<SidebarProps> = ({
  items,
  activeItemId,
  onItemClick,
}) => {
  const { t } = useTranslation();

  const handleOnClick = (e: React.MouseEvent, action: () => void) => {
    e.preventDefault();
    action();
  };

  return (
    <SideNav
      isFixedNav
      expanded
      className={`${styles.sidebar}`}
      aria-label="Side Navigation"
      data-testid="sidebar"
    >
      <SideNavItems>
        {items.map((item) => (
          <SideNavLink
            key={item.id}
            renderIcon={() => (
              <BahmniIcon
                name={item.icon}
                id={`sidebar-icon-${item.id}`}
                size={ICON_SIZE.SM}
              />
            )}
            href="#"
            onClick={(e) => handleOnClick(e, () => onItemClick(item.id))}
            isActive={item.id === activeItemId}
            data-testid={`sidebar-item-${item.id}`}
          >
            {t(item.label)}
          </SideNavLink>
        ))}
      </SideNavItems>
    </SideNav>
  );
};

export default Sidebar;
