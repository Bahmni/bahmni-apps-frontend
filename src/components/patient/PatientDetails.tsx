import React, { useMemo } from 'react';
import { usePatient } from '@hooks/usePatient';
import { InlineNotification, SkeletonText, Tile, Stack } from '@carbon/react';
import { formatPatientData } from '@services/patientService';

interface PatientDetailsProps {
  patientUUID: string;
}

const PatientDetails: React.FC<PatientDetailsProps> = ({ patientUUID }) => {
  const { patient, loading, error } = usePatient(patientUUID);

  // Format patient data using the service
  const formattedPatient = useMemo(() => {
    if (!patient) return null;
    return formatPatientData(patient);
  }, [patient]);

  if (loading) {
    return (
      <Tile>
        <SkeletonText
          heading
          width="100%"
          lineCount={5}
          data-testid="skeleton-loader"
        />
      </Tile>
    );
  }

  if (error) {
    return (
      <InlineNotification kind="error" title="Error" subtitle={error.message} />
    );
  }

  if (!patient || !formattedPatient) {
    return (
      <InlineNotification
        kind="info"
        title="No data"
        subtitle="Patient information not found"
      />
    );
  }

  return (
    <Stack gap={5}>
      <Tile>
        {formattedPatient.fullName && <h2>{formattedPatient.fullName}</h2>}
        <p>ID: {formattedPatient.id}</p>
        {formattedPatient.gender && <p>Gender: {formattedPatient.gender}</p>}
        {formattedPatient.birthDate && (
          <p>Birth Date: {formattedPatient.birthDate}</p>
        )}
        {formattedPatient.formattedAddress && (
          <p>Address: {formattedPatient.formattedAddress}</p>
        )}
        {formattedPatient.formattedContact && (
          <p>Contact: {formattedPatient.formattedContact}</p>
        )}
      </Tile>
    </Stack>
  );
};

export default PatientDetails;
