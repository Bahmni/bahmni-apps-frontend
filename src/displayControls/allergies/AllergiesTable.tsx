import React, { useMemo } from 'react';
import { Tag } from '@carbon/react';
import { useTranslation } from 'react-i18next';
import { ExpandableDataTable } from '@components/common/expandableDataTable/ExpandableDataTable';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { useAllergies } from '@hooks/useAllergies';
import { formatAllergies } from '@services/allergyService';
import { FormattedAllergy } from '@types/allergy';
import { formatDateTime } from '@utils/date';
import { generateId, capitalize } from '@utils/common';

/**
 * Component to display patient allergies in a DataTable with expandable rows
 */
const AllergiesTable: React.FC = () => {
  const { t } = useTranslation();
  const patientUUID = usePatientUUID();
  const { allergies, loading, error } = useAllergies(patientUUID);

  // Define table headers
  const headers = useMemo(
    () => [
      { key: 'display', header: t('ALLERGEN') },
      { key: 'severity', header: t('SEVERITY') },
      { key: 'manifestation', header: t('REACTIONS') },
      { key: 'status', header: t('ALLERGY_LIST_STATUS') },
      { key: 'recorder', header: t('ALLERGY_LIST_PROVIDER') },
      { key: 'recordedDate', header: t('ALLERGY_LIST_RECORDED_DATE') },
    ],
    [t],
  );

  const sortable = useMemo(
    () => [
      { key: 'display', sortable: true },
      { key: 'severity', sortable: true },
      { key: 'manifestation', sortable: false },
      { key: 'status', sortable: true },
      { key: 'recorder', sortable: true },
      { key: 'recordedDate', sortable: true },
    ],
    [],
  );

  // Format allergies for display
  const formattedAllergies = useMemo(() => {
    if (!allergies || allergies.length === 0) return [];
    return formatAllergies(allergies);
  }, [allergies]);

  // Create row class names array for styling rows with severe allergies
  const rowClassNames = useMemo(() => {
    const classNames: Record<string, string> = {};

    formattedAllergies.forEach((allergy) => {
      if (allergy.id && allergy.severity && allergy.severity === 'severe') {
        classNames[allergy.id] = 'criticalCell';
      }
    });

    return classNames;
  }, [formattedAllergies]);

  // Function to render cell content based on the cell ID
  const renderCell = (allergy: FormattedAllergy, cellId: string) => {
    switch (cellId) {
      case 'display':
        return allergy.display;
      case 'status':
        return (
          <Tag type={allergy.status === 'Active' ? 'green' : 'gray'}>
            {allergy.status === 'Active'
              ? t('ALLERGY_LIST_ACTIVE')
              : t('ALLERGY_LIST_INACTIVE')}
          </Tag>
        );
      case 'manifestation':
        return allergy.reactions
          ? allergy.reactions
              .map((reaction) => reaction.manifestation.join(', '))
              .join(', ')
          : t('ALLERGY_TABLE_NOT_AVAILABLE');
      case 'severity':
        return capitalize(allergy.severity || 'Unknown');
      case 'recorder':
        return allergy.recorder || t('ALLERGY_TABLE_NOT_AVAILABLE');
      case 'recordedDate': {
        const recordedDate = formatDateTime(allergy.recordedDate || '');
        return recordedDate.formattedResult || t('ALLERGY_TABLE_NOT_AVAILABLE');
      }
    }
  };

  // Function to render expanded content for an allergy
  const renderExpandedContent = (allergy: FormattedAllergy) => {
    if (allergy.note && allergy.note.length > 0) {
      return (
        <p style={{ padding: '0.5rem' }} key={generateId()}>
          {allergy.note.join(', ')}
        </p>
      );
    }
    return undefined;
  };

  return (
    <div data-testid="allergy-table">
      <ExpandableDataTable
        tableTitle={t('ALLERGIES_DISPLAY_CONTROL_HEADING')}
        rows={formattedAllergies}
        headers={headers}
        sortable={sortable}
        renderCell={renderCell}
        renderExpandedContent={renderExpandedContent}
        loading={loading}
        error={error}
        ariaLabel={t('ALLERGIES_DISPLAY_CONTROL_HEADING')}
        emptyStateMessage={t('NO_ALLERGIES')}
        rowClassNames={rowClassNames}
      />
    </div>
  );
};

export default AllergiesTable;
