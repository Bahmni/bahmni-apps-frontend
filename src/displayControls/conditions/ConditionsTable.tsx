import React, { useMemo } from 'react';
import { Tag } from '@carbon/react';
import { useTranslation } from 'react-i18next';
import { ExpandableDataTable } from '@components/common/expandableDataTable/ExpandableDataTable';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { useConditions } from '@hooks/useConditions';
import { formatConditions } from '@services/conditionService';
import { ConditionStatus, FormattedCondition } from '@types/condition';
import { formatDate, formatDateTime } from '@utils/date';
import { FormatDateResult } from '@types/date';

/**
 * Maps condition status to appropriate tag type
 * @param status - The condition status
 * @returns The tag type for the status
 */
const getStatusTagType = (status: ConditionStatus): 'green' | 'gray' => {
  switch (status) {
    case ConditionStatus.Active:
      return 'green';
    case ConditionStatus.Inactive:
    default:
      return 'gray';
  }
};

/**
 * Component to display patient conditions in a DataTable with expandable rows
 */
const ConditionsTable: React.FC = () => {
  const { t } = useTranslation();
  const patientUUID = usePatientUUID();
  const { conditions, loading, error } = useConditions(patientUUID);

  // Define table headers
  const headers = useMemo(
    () => [
      { key: 'display', header: t('CONDITION_LIST_CONDITION') },
      { key: 'status', header: t('CONDITION_LIST_STATUS') },
      { key: 'onsetDate', header: t('CONDITION_TABLE_ONSET_DATE') },
      { key: 'recorder', header: t('CONDITION_TABLE_PROVIDER') },
      { key: 'recordedDate', header: t('CONDITION_TABLE_RECORDED_DATE') },
    ],
    [t],
  );

  // Format conditions for display
  const formattedConditions = useMemo(() => {
    if (!conditions || conditions.length === 0) return [];
    return formatConditions(conditions);
  }, [conditions]);

  // Function to render cell content based on the cell ID
  const renderCell = (condition: FormattedCondition, cellId: string) => {
    switch (cellId) {
      case 'display':
        return condition.display;
      case 'status':
        return (
          <Tag type={getStatusTagType(condition.status)}>
            {condition.status == ConditionStatus.Active
              ? t('CONDITION_LIST_ACTIVE')
              : t('CONDITION_LIST_INACTIVE')}
          </Tag>
        );
      case 'onsetDate': {
        const onsetDate: FormatDateResult = formatDate(
          condition.onsetDate || '',
        );
        return onsetDate.formattedResult || t('CONDITION_TABLE_NOT_AVAILABLE');
      }
      case 'recorder':
        return condition.recorder || t('CONDITION_TABLE_NOT_AVAILABLE');
      case 'recordedDate': {
        const recordedDate: FormatDateResult = formatDateTime(
          condition.recordedDate || '',
        );
        return (
          recordedDate.formattedResult || t('CONDITION_TABLE_NOT_AVAILABLE')
        );
      }
    }
  };

  // Function to render expanded content for a condition
  const renderExpandedContent = (condition: FormattedCondition) => {
    if (condition.note && condition.note.length > 0) {
      return condition.note.map((note, index) => (
        <p style={{ padding: '0.5rem' }} key={index}>
          {note}
        </p>
      ));
    }
    return undefined;
  };

  return (
    <div data-testid="condition-table">
      <ExpandableDataTable
        tableTitle={t('CONDITION_LIST_DISPLAY_CONTROL_TITLE')}
        rows={formattedConditions}
        headers={headers}
        renderCell={renderCell}
        renderExpandedContent={renderExpandedContent}
        loading={loading}
        error={error}
        ariaLabel={t('CONDITION_LIST_DISPLAY_CONTROL_TITLE')}
        emptyStateMessage={t('CONDITION_LIST_NO_CONDITIONS')}
      />
    </div>
  );
};

export default ConditionsTable;
