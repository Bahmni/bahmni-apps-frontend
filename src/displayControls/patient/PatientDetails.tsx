import { SkeletonText } from '@carbon/react';
import React, { useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import BahmniIcon from '@/components/common/bahmniIcon/BahmniIcon';
import { ICON_SIZE } from '@/constants/icon';
import { usePatient } from '@hooks/usePatient';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { formatPatientData } from '@services/patientService';
import * as styles from './__styles__/PatientDetails.module.scss';

// TODO: Extract this as a PatientDetails Display Control Component
const PatientDetails: React.FC = () => {
  const { t } = useTranslation();
  const patientUUID: string | null = usePatientUUID();
  const { patient, loading, error } = usePatient(patientUUID);

  // Format patient data using the service
  const formattedPatient = useMemo(() => {
    if (!patient) return null;
    return formatPatientData(patient);
  }, [patient]);

  if (loading || error || !patient || !formattedPatient) {
    return (
      <div className={styles.skeletonContainer}>
        <SkeletonText
          heading
          width="20%"
          lineCount={2}
          data-testid="skeleton-loader"
        />
        <SkeletonText
          width="50%"
          lineCount={3}
          data-testid="skeleton-loade-subheader"
        />
      </div>
    );
  }

  const formatField = (value?: string | number | null) => value ?? null;

  const formattedIdentifiers = formattedPatient.identifiers.size
    ? Array.from(formattedPatient.identifiers.entries())
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        .map(([key, value]) => `${value}`)
        .filter(Boolean)
        .join(' | ')
    : null;

  const formattedGender = formatField(formattedPatient.gender);

  const formattedAge =
    formattedPatient.age?.years !== undefined
      ? `${formattedPatient.age.years} ${t('CLINICAL_YEARS_TRANSLATION_KEY', { count: formattedPatient.age.years })}, ${formattedPatient.age.months} ${t('CLINICAL_MONTHS_TRANSLATION_KEY', { count: formattedPatient.age.months })}, ${formattedPatient.age.days} ${t('CLINICAL_DAYS_TRANSLATION_KEY', { count: formattedPatient.age.days })}`
      : null;

  const details = [formattedAge, formatField(formattedPatient.birthDate)]
    .filter(Boolean)
    .join(' | ');

  return (
    <div className={styles.header}>
      {formattedPatient.fullName && (
        <p data-testid="patient-name" className={styles.patientName}>
          {formattedPatient.fullName}
        </p>
      )}
      <div className={styles.patientDetails}>
        <div className={styles.identifierAndGenderWrapper}>
          {formattedIdentifiers && (
            <p className={styles.detailsWithIcon}>
              <BahmniIcon id="id-card" name="fa-id-card" size={ICON_SIZE.SM} />
              <span>{formattedIdentifiers}</span>
            </p>
          )}
          {formattedGender && (
            <p className={styles.detailsWithIcon}>
              <BahmniIcon
                id="gender"
                name="fa-mars-stroke-up"
                size={ICON_SIZE.SM}
              />
              <span>{formattedGender}</span>
            </p>
          )}
        </div>
        {details && (
          <p className={styles.detailsWithIcon}>
            <BahmniIcon id="age" name="fa-cake-candles" size={ICON_SIZE.SM} />
            <span>{details}</span>
          </p>
        )}
      </div>
    </div>
  );
};

export default PatientDetails;
