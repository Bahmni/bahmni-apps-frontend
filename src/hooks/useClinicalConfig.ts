import { useContext } from 'react';
import { ClinicalConfigContext } from '@contexts/ClinicalConfigContext';
import { ClinicalConfigContextType } from '@types/config';

/**
 * Custom hook to access the config context
 * @returns The config context values including config, loading state, and error
 */
export const useClinicalConfig = (): ClinicalConfigContextType => {
  const context = useContext(ClinicalConfigContext);

  if (!context) {
    throw new Error(
      'useClinicalConfig must be used within a ClinicalConfigProvider',
    );
  }

  return context;
};
