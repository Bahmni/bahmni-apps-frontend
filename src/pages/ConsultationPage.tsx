import React, { Suspense } from 'react';
import { Grid, Column, Section, Loading } from '@carbon/react';
import ClinicalLayout from '@layouts/clinical/ClinicalLayout';
import PatientDetails from '@displayControls/patient/PatientDetails';
import ConditionsTable from '@displayControls/conditions/ConditionsTable';
import AllergiesTable from '@displayControls/allergies/AllergiesTable';
import { useClinicalConfig } from '@hooks/useClinicalConfig';
import Header from '@components/clinical/header/Header';
import Sidebar, { SidebarItemProps } from '@components/common/sidebar/Sidebar';
import { Dashboard } from '@types/config';
import { useDashboardConfig } from '@hooks/useDashboardConfig';
import { DashboardConfig } from '@types/dashboardConfig';
import useNotification from '@hooks/useNotification';

const ConsultationPage: React.FC = () => {
  const { clinicalConfig } = useClinicalConfig();
  const { addNotification } = useNotification();

  const getDefaultDashboard = (dashboards: Dashboard[]): Dashboard | null => {
    const defaultDashboard = dashboards.find(
      (dashboard) => dashboard.default === true,
    );
    if (defaultDashboard) {
      return defaultDashboard;
    }
    return null;
  };

  const getSidebarItems = (
    dashboardConfig: DashboardConfig,
  ): SidebarItemProps[] => {
    return dashboardConfig.sections.map((section) => ({
      id: section.name,
      icon: section.icon,
      // TODO: add translation
      label: section.name,
      active: false,
      action: () => {},
    }));
  };

  if (!clinicalConfig) {
    return <Loading description="Loading..." />;
  }
  const currentDashboard = getDefaultDashboard(clinicalConfig.dashboards);

  if (!currentDashboard) {
    addNotification({
      title: 'Error',
      message: 'No default dashboard configured',
      type: 'error',
    });
    return <Loading description="Error Loading dashboard" />;
  }
  const { dashboardConfig } = useDashboardConfig(currentDashboard.url);

  if (!dashboardConfig) {
    return <Loading description="Loading dashboard config..." />;
  }

  return (
    <ClinicalLayout
      header={<Header />}
      patientDetails={<PatientDetails />}
      sidebar={<Sidebar items={getSidebarItems(dashboardConfig)} />}
      mainDisplay={
        <Suspense fallback="loading">
          <Section>
            <Grid>
              <Column lg={16} md={8} sm={4}>
                <AllergiesTable />
                <ConditionsTable />
              </Column>
            </Grid>
          </Section>
        </Suspense>
      }
    />
  );
};

export default ConsultationPage;
