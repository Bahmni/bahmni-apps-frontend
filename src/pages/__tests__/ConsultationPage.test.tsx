import React, { ReactNode } from 'react';
import { render, screen } from '@testing-library/react';
import ConsultationPage from '../ConsultationPage';
import PatientDetails from '@displayControls/patient/PatientDetails';
import ConditionsTable from '@displayControls/conditions/ConditionsTable';
import AllergiesTable from '@displayControls/allergies/AllergiesTable';
import { useClinicalConfig } from '@hooks/useClinicalConfig';
import { useDashboardConfig } from '@hooks/useDashboardConfig';
import useNotification from '@hooks/useNotification';
import {
  validFullClinicalConfig,
  validDashboardConfig,
} from '@__mocks__/configMocks';
import Sidebar, { SidebarItemProps } from '@components/common/sidebar/Sidebar';

// Mock React.Suspense to render children immediately in tests
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  Suspense: ({ children }: { children: ReactNode }) => children,
}));

// Mock Carbon components
jest.mock('@carbon/react', () => ({
  Grid: jest.fn(({ children }) => (
    <div data-testid="carbon-grid">{children}</div>
  )),
  Column: jest.fn(({ children }) => (
    <div data-testid="carbon-column">{children}</div>
  )),
  Section: jest.fn(({ children }) => (
    <div data-testid="carbon-section">{children}</div>
  )),
  Loading: jest.fn(() => <div data-testid="carbon-loading">Loading...</div>),
}));

jest.mock('@layouts/clinical/ClinicalLayout', () => {
  return jest.fn(({ header, patientDetails, sidebar, mainDisplay }) => (
    <div data-testid="mocked-clinical-layout">
      <div data-testid="mocked-header">{header}</div>
      <div data-testid="mocked-patient-section">{patientDetails}</div>
      <div data-testid="mocked-sidebar">{sidebar}</div>
      <div data-testid="mocked-main-display">{mainDisplay}</div>
    </div>
  ));
});

// Mock the Header component
jest.mock('@components/clinical/header/Header', () => {
  return jest.fn(() => (
    <div data-testid="mocked-header-component">Mocked Header</div>
  ));
});

// Mock the PatientDetails component
jest.mock('@displayControls/patient/PatientDetails', () => {
  return jest.fn(() => (
    <div data-testid="mocked-patient-details">Mocked PatientDetails</div>
  ));
});

// Mock the ConditionsTable component
jest.mock('@displayControls/conditions/ConditionsTable', () => {
  return jest.fn(() => (
    <div data-testid="mocked-conditions-table">Mocked ConditionsTable</div>
  ));
});

// Mock the useClinicalConfig hook
jest.mock('@hooks/useClinicalConfig');

// Mock the useDashboardConfig hook
jest.mock('@hooks/useDashboardConfig');

// Mock the useNotification hook
jest.mock('@hooks/useNotification');

// Mock the AllergiesTable component
jest.mock('@displayControls/allergies/AllergiesTable', () => {
  return jest.fn(() => (
    <div data-testid="mocked-allergy-table">Mocked AllergiesTable</div>
  ));
});

// Mock the Sidebar component
jest.mock('@components/common/sidebar/Sidebar', () => {
  return jest.fn(({ items }: { items: SidebarItemProps[] }) => (
    <div data-testid="mocked-sidebar-component">
      {items.map((item: SidebarItemProps) => (
        <div key={item.id} data-testid={`sidebar-item-${item.id}`}>
          {item.label}
        </div>
      ))}
    </div>
  ));
});
describe('ConsultationPage Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(console, 'log').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation(() => {});

    // Default mock for useNotification
    (useNotification as jest.Mock).mockReturnValue({
      addNotification: jest.fn(),
    });
  });

  it('should show loading state when clinical config is null', () => {
    // Mock useClinicalConfig to return null config (loading state)
    (useClinicalConfig as jest.Mock).mockReturnValue({ clinicalConfig: null });

    render(<ConsultationPage />);

    // Should show loading component
    expect(screen.getByTestId('carbon-loading')).toBeInTheDocument();

    // Should not render other components
    expect(
      screen.queryByTestId('mocked-patient-details'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('mocked-conditions-table'),
    ).not.toBeInTheDocument();
  });

  it('should show error notification when no default dashboard is configured', () => {
    // Mock a clinical config with no default dashboard
    const configWithNoDefault = {
      ...validFullClinicalConfig,
      dashboards: validFullClinicalConfig.dashboards.map((dashboard) => ({
        ...dashboard,
        default: false,
      })),
    };

    // Mock useClinicalConfig to return config with no default dashboard
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: configWithNoDefault,
    });

    const mockAddNotification = jest.fn();
    (useNotification as jest.Mock).mockReturnValue({
      addNotification: mockAddNotification,
    });

    render(<ConsultationPage />);

    // Should show loading component with error message
    expect(screen.getByTestId('carbon-loading')).toBeInTheDocument();

    // Should call addNotification with error message
    expect(mockAddNotification).toHaveBeenCalledWith({
      title: 'Error',
      message: 'No default dashboard configured',
      type: 'error',
    });
  });

  it('should show loading state when dashboard config is being loaded', () => {
    // Mock useClinicalConfig to return config with default dashboard
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig to return null (loading state)
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: null,
      isLoading: true,
      error: null,
    });

    render(<ConsultationPage />);

    // Should show loading component for dashboard config
    expect(screen.getByTestId('carbon-loading')).toBeInTheDocument();
  });

  it('should render with correct Carbon layout structure when all configs are loaded', async () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig to return dashboard config
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: validDashboardConfig,
      isLoading: false,
      error: null,
    });

    render(<ConsultationPage />);

    // Should render Carbon layout components
    expect(await screen.findByTestId('carbon-section')).toBeInTheDocument();
    expect(await screen.findByTestId('carbon-grid')).toBeInTheDocument();
    expect(await screen.findByTestId('carbon-column')).toBeInTheDocument();
  });

  it('should render child components when all configs are loaded', async () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig to return dashboard config
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: validDashboardConfig,
      isLoading: false,
      error: null,
    });

    render(<ConsultationPage />);

    // Should render child components
    expect(PatientDetails).toHaveBeenCalled();
    expect(
      await screen.findByTestId('mocked-patient-details'),
    ).toBeInTheDocument();

    expect(ConditionsTable).toHaveBeenCalled();
    expect(AllergiesTable).toHaveBeenCalled();
    expect(
      await screen.findByTestId('mocked-conditions-table'),
    ).toBeInTheDocument();
    expect(
      await screen.findByTestId('mocked-allergy-table'),
    ).toBeInTheDocument();
  });

  it('should generate sidebar items correctly from dashboard config', async () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig to return dashboard config
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: validDashboardConfig,
      isLoading: false,
      error: null,
    });

    render(<ConsultationPage />);

    // Should render sidebar with correct items
    expect(Sidebar).toHaveBeenCalled();
    expect(screen.getByTestId('mocked-sidebar-component')).toBeInTheDocument();

    // Check if sidebar items are rendered correctly
    expect(screen.getByTestId('sidebar-item-Vitals')).toBeInTheDocument();
    expect(screen.getByTestId('sidebar-item-Medications')).toBeInTheDocument();

    // Verify sidebar props
    const sidebarCall = (Sidebar as jest.Mock).mock.calls[0][0];
    expect(sidebarCall.items).toHaveLength(2);
    expect(sidebarCall.items[0]).toEqual({
      id: 'Vitals',
      icon: 'heartbeat',
      label: 'Vitals',
      active: false,
      action: expect.any(Function),
    });
    expect(sidebarCall.items[1]).toEqual({
      id: 'Medications',
      icon: 'pills',
      label: 'Medications',
      active: false,
      action: expect.any(Function),
    });
  });

  it('should call useDashboardConfig with correct dashboard URL', () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: validDashboardConfig,
      isLoading: false,
      error: null,
    });

    render(<ConsultationPage />);

    // Get the default dashboard URL from validFullClinicalConfig
    const defaultDashboard = validFullClinicalConfig.dashboards.find(
      (dashboard) => dashboard.default === true,
    );

    // Verify useDashboardConfig was called with correct URL
    expect(useDashboardConfig).toHaveBeenCalledWith(defaultDashboard?.url);
  });

  it('should match snapshot when loading clinical config', () => {
    // Mock useClinicalConfig to return null config (loading state)
    (useClinicalConfig as jest.Mock).mockReturnValue({ clinicalConfig: null });

    const { asFragment } = render(<ConsultationPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when loading dashboard config', () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig to return null (loading state)
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: null,
      isLoading: true,
      error: null,
    });

    const { asFragment } = render(<ConsultationPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when fully loaded', () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    // Mock useDashboardConfig to return dashboard config
    (useDashboardConfig as jest.Mock).mockReturnValue({
      dashboardConfig: validDashboardConfig,
      isLoading: false,
      error: null,
    });

    const { asFragment } = render(<ConsultationPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
