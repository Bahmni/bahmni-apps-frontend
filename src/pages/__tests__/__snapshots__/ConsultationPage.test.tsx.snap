// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsultationPage Rendering and Structure should match the snapshot 1`] = `
<div>
  <div
    data-testid="mocked-clinical-layout"
  >
    <div
      data-testid="mocked-header"
    >
      <div
        data-testid="mocked-header-component"
      >
        <div
          data-testid="sidenav-item-vitals"
        >
          Vitals
        </div>
        <div
          data-testid="sidenav-item-medications"
        >
          Medications
        </div>
        <div
          data-testid="active-sidenav-item"
        >
          Vitals
        </div>
      </div>
    </div>
    <div
      data-testid="mocked-patient-section"
    >
      <div
        data-testid="carbon-tile"
      >
        <div
          data-testid="carbon-grid"
        >
          <div
            data-testid="carbon-column"
          >
            <div
              data-testid="mocked-patient-details"
            >
              Mocked PatientDetails
            </div>
          </div>
          <div
            data-testid="carbon-column"
          >
            <button
              data-testid="carbon-button"
            >
              translated_PATIENT_HEADER_SHOW_ACTION_AREA
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="mocked-sidebar"
    />
    <div
      data-testid="mocked-main-display"
    >
      <div
        data-testid="mocked-dashboard-container"
      >
        <div
          data-testid="dashboard-sections-count"
        >
          2
        </div>
        <div
          data-testid="dashboard-active-item"
        >
          Vitals
        </div>
      </div>
    </div>
  </div>
</div>
`;
